Modules\Vlans

Npeca75 opened the pull request:
removed all ports without ifIndex pair in `ports` table

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

Npeca75 commented:
vsolution_v1600d.json

```
array (
  1 => 
  array (
    'ifDescr' => '',
    'ifName' => 'GE0/1',
    'ifAlias' => '',
    'ifType' => 'ethernetCsmacd',
    'ifOperStatus' => 'down',
```
have empty descr for first 6 interfaces, so there is no ifIndex where to attach vlans

Npeca75 commented:
voss_8608.json

broken BRIDGE-MIB::dot1dBasePortIfIndex

```
dot1dBasePortIfIndex[192] = 0
dot1dBasePortIfIndex[196] = 0
dot1dBasePortIfIndex[200] = 0
dot1dBasePortIfIndex[204] = 0
dot1dBasePortIfIndex[208] = 0
dot1dBasePortIfIndex[209] = 0
dot1dBasePortIfIndex[210] = 0
dot1dBasePortIfIndex[211] = 0
dot1dBasePortIfIndex[212] = 0
dot1dBasePortIfIndex[213] = 0
dot1dBasePortIfIndex[214] = 0
dot1dBasePortIfIndex[215] = 0
dot1dBasePortIfIndex[216] = 0
dot1dBasePortIfIndex[217] = 0
dot1dBasePortIfIndex[218] = 0
dot1dBasePortIfIndex[219] = 0
```

Npeca75 commented:
nxos_n3k-3064pq.json

baseport = 0
no matching interfaces

Npeca75 commented:
linksys-ss_lgs318p.json

baseport = 1003
same as **RADLAN**

Npeca75 commented:
eltex-mes23xx_mes2324b.json

"baseport": 163,
no such port

Npeca75 commented:
ios_c3560

no such port `"baseport": 33,`

```
dot1dBasePortIfIndex[32] = 10028
dot1dBasePortIfIndex[34] = 10030
```

Npeca75 commented:
aos.json

same as **RADLAN**

"baseport": 1003,

murrant commented on LibreNMS/Interfaces/Discovery/VlanDiscovery.php:
```diff
@@ -0,0 +1,14 @@
+{
+    /**
+     * Discover processors.
+     * Returns an array of LibreNMS\App\Vlan && LibreNMS\App\PortVlan objects
```
Why are there two types of data mixed together?  Perhaps there should be two discovery methods?
On newer modules, I've been returning collections of Models to help keep the data more consistent.

murrant replied:
Something like this, what do you think?

discoverVlans(): Collection<\App\Models\Vlan>;
discoverVlanPorts(): Collection<\App\Models\PortVlan>;

Npeca75 replied:
> discoverVlans(): Collection<\App\Models\Vlan>;
> discoverVlanPorts(Collection $vlans): Collection<\App\Models\PortVlan>;

hi @murrant 
correct me if i am wrong ...
my intention was to put whole vlan discovery in one function
since it have 
[basic] data, name,type
and
[ports] data, tagged,portid,etc
i needed to return somehow this dual data, and array was handy
one function, one array,then split & process at centralized place

but, if you insist to split them in two function, discoverBasicVlan, discoverPortVlan ....

how to be then ?

murrant replied:
I'm just trying to make things match our other code so they are more consistent. No discovery interfaces we currently have return multiple sets of data.

The data seems unconnected to me so it seems separate functions makes sense.  You've gone through all the code though, so you might have more context to me.

If we do want to return both, I don't like using arrays for that I would much prefer a data transfer object to hold the data in a structured way that we can expect.  But again this would be unique, different from every other module we currently have.

murrant commented on LibreNMS/Modules/Vlans.php:
```diff
@@ -0,0 +1,264 @@
+        $this->syncModels($os->getDevice(), 'vlans', $vlans);
+
+        foreach ($vlanData['ports'] ?? [] as $key => $data) {
+            $data['vlan'] = $data['vlan'] ?? 0;
```
If these were already PortVlan object, we wouldn't need all these null checks here.

Npeca75 commented:
**ios_stp-vlans**

there is no vlan98 in test data

murrant commented:
@Npeca75 Actually, there is, it is in ios_stp-vlans@98.snmprec  This test ensures the context based polling is working.

FYI, I've been working on this a bit locally and will push when my changes aren't totally broken :)  I'll admit the split discovery seems a little odd.  it is basically, vlans with a name mapping vs vlans with a port mapping.

murrant commented:
@Npeca75 I was looking at ao6 and ao7 and the test data for aos6 seems to include Q-BRIDGE-MIB data, but not ALCATEL-IND1-VLAN-MGR-MIB DEFINITIONS. 

It makes me wonder if we even need the vendor mibs.  Any idea?

murrant commented:
@Npeca75 if tests pass, this looks good to me now.  Do you have any objections/changes now that I changed a lot?

Npeca75 commented:
> @Npeca75 if tests pass, this looks good to me now. Do you have any objections/changes now that I changed a lot?

tnx @murrant for your support
i see you made a lot of cleanup & tidy-up :)
if the tests pass, let's merge it

Npeca75 commented:
ahh, i see

`LibreNMS/OS/Traits/BridgeMib.php`

so we have now centralized function for 
`BRIDGE-MIB::dot1dBasePortIfIndex`

need to rewrite FDB and LINKS ...

murrant commented:
Yeah, I settled on BridgeMib trait and I think that is a good place for it.

Npeca75 commented:
@murrant 

about `[$baseport - 1] ?? 0,`
it was in old PHP code, since i have no `boss` device at my hand, i don't dare to change this `-1` modifier

murrant commented:
Needs #18040 

That was a fun one to track down.  relationship was loaded empty, then not refreshed because it was already loaded even though vlans already were updated.  Setting the relation after syncing make sure the data is up-to-date and saves db round trips.

Npeca75 commented:
> hmmm @murrant something is wrong with `junos_rpm` ... i could not produce consistent test data

ok @murrant 

i was able to find the error :)
vlan_id instead of vlan_vlan
and since junos looping through vlan_domains, vlan_is was incementing always and every junos_rpm test failed

fixed

murrant commented:
@Npeca75 hmm, maybe setting "port group" to  domain was a bad idea?

Npeca75 commented:
hi @murrant 

timeouts on 8.2 test was here from first day, so no new errors are introduced
simply timeouts on OS where we have multiple context

murrant commented:
So, we're screwed until we can speed up tests? I don't think we can increase the timeout for github actions.

Npeca75 commented:
> testing_config.yml already contains: snmp.timeout: 0.1 snmp.retries: 0

i am out of idea
if you put timeouts and retries too low, then (my opinion) there is no time left for snmpsim context@98, at least, last error state is about vlan98
if you put it high enough, then there will be general timeout

i tried to set workflow timeout to 3hrs
but failing vlan98 prevent this modification

Npeca75 commented:
hi @murrant 

i restored the order of data collection in VRP OS
first vendor, then if empty, standard q-bridge ...

at least, on my 
```
HardwareS5300
Operating SystemHuawei VRP 5.110 (V200R001C00SPC300) [V200R001SPH005]
```

standard q-bridge MIBs have incomplete data, so it is useless without vendor MIBs

murrant commented:
Yeah, that is why I refactored the ordering so we could override it per-OS.  I just didn't get to the part where I figured out what order each OS needed.

There is still something messed up with this.  The long test indicates we could cause a lot of delay in real environments too.  I hope to go through and evaluate if we can minimize trying to discover non-existent vlans more.

Npeca75 commented:
> There is still something messed up with this

hi @murrant 

i put a time measurement in vlan processing and, there is nothing wrong with times

only 2-3 OS have longer processing time
bdcom - 90 sec
fs-bdcom - 70 sec
nxos - 25 sec

but this is only few wasted minutes, i don't expect any change in final test timeout even if we fix this cca 4-5 minute

murrant commented:
Why is discovery run twice?

2025-07-17T13:15:22.3145834Z ."START" // LibreNMS/Modules/Vlans.php:75
2025-07-17T13:15:22.3146574Z "fs-bdcom" // LibreNMS/Modules/Vlans.php:75
2025-07-17T13:15:22.3147203Z "S3900-24F4S-R" // LibreNMS/Modules/Vlans.php:75
2025-07-17T13:15:24.9306399Z "Basic Vlan discovery: 2" // LibreNMS/Modules/Vlans.php:90
2025-07-17T13:16:48.1726914Z "Port Vlan discovery: 84" // LibreNMS/Modules/Vlans.php:105
2025-07-17T13:16:49.8252231Z "START" // LibreNMS/Modules/Vlans.php:75
2025-07-17T13:16:49.8252899Z "fs-bdcom" // LibreNMS/Modules/Vlans.php:75
2025-07-17T13:16:49.8253518Z "S3900-24F4S-R" // LibreNMS/Modules/Vlans.php:75
2025-07-17T13:16:50.6948403Z "Basic Vlan discovery: 1" // LibreNMS/Modules/Vlans.php:90
2025-07-17T13:18:03.6007119Z "Port Vlan discovery: 73" // LibreNMS/Modules/Vlans.php:105

murrant commented:
Ok, so it is BDCOM that is messing us up... It is insane.  Every VLAN * 1057 ports.  This is making 4 million port entries.

Npeca75 commented:
> Why is discovery run twice?

ok @murrant 

i reverted back to 

https://github.com/librenms/librenms/pull/17870/commits/1b22c93cdc0c67c80c1139539da36b5a81af109a

where i split discovery to two function
there is NO duplicate discovery

so ... something after that commit

murrant commented:
@Npeca75 bdcom with 4095*1057 vlan ports was 100% the problem.  It running twice didn't help :)

Npeca75 commented:
> @Npeca75 bdcom with 4095*1057 vlan ports was 100% the problem. It running twice didn't help :)

ok, wuld you please finish this PR then?
put back poller,remove time() debug ...etc
to avoid duplicate pushing

murrant reviewed (approved):
Looking good.  I think we even reduced the CI test time :)
